shared/src/commonMain/kotlin/org/hyperskill/app/profile/view/BadgesViewStateMapper.kt

```diff
@@ -49,44 +50,159 @@ class BadgesViewStateMapper(
         )
 
     private fun mapLockedBadge(badgeKind: BadgeKind): BadgesViewState.BadgeViewState =
         BadgesViewState.BadgeViewState(
             kind = badgeKind,
-            title = getBadgeTitle(badgeKind),
+            title = getBadgeTitle(badgeKind) ?: "",
             image = BadgesViewState.BadgeImage.Locked,
             formattedCurrentLevel = resourceProvider.getString(SharedResources.strings.badge_level, 0),
             nextLevel = 1,
             progress = 0f
         )
 
-    private fun getBadgeTitle(badgeKind: BadgeKind): String =
+    fun map(badgeDetails: ProfileFeature.Action.ViewAction.BadgeDetails): BadgeDetailsViewState =
+        when (badgeDetails) {
+            is ProfileFeature.Action.ViewAction.BadgeDetails.Badge -> mapToDetails(badgeDetails.badge)
+            is ProfileFeature.Action.ViewAction.BadgeDetails.Kind -> mapToDetails(badgeDetails.badgeKind)
+        }
+
+    private fun mapToDetails(badge: Badge): BadgeDetailsViewState =
+        BadgeDetailsViewState(
+            kind = badge.kind,
+            rank = badge.rank,
+            title = badge.title,
+            formattedRank = getBadgeRankName(badge.rank) ?: "",
+            badgeDescription = if (!badge.isMaxLevel) {
+                getBadgeDescription(badge.kind)
+            } else {
+                null
+            },
+            levelDescription = when {
+                badge.level == 0 ->
+                    badge.nextLevelValue?.let { nextLevelValue ->
+                        getBadgeUnlockDescription(badge.kind, nextLevelValue)
+                    }
+                !badge.isMaxLevel ->
+                    badge.nextLevelValue?.let { nextLevelValue ->
+                        getBadgeUpgradeDescription(badge.kind, nextLevelValue)
+                    }
+                badge.isMaxLevel ->
+                    resourceProvider.getString(SharedResources.strings.badge_max_level_description)
+                else -> null
+            },
+            formattedCurrentLevel = if (badge.level == 0) {
+                resourceProvider.getString(SharedResources.strings.badge_locked)
+            } else {
+                resourceProvider.getString(SharedResources.strings.badge_level, badge.level)
+            },
+            formattedNextLevel = getNextLevel(badge)?.let { nextLevel ->
+                resourceProvider.getString(SharedResources.strings.badge_level, nextLevel)
+            },
+            progress = getProgress(badge),
+            imageSource = badge.imageFull,
+            isLocked = false // TODO: replace with `badge.rank == BadgeRank.LOCKED`
+        )
+
+    private fun mapToDetails(badgeKind: BadgeKind): BadgeDetailsViewState =
+        BadgeDetailsViewState(
+            kind = badgeKind,
+            rank = BadgeRank.UNKNOWN, // TODO: replace with BadgeRank.LOCKED
+            title = getBadgeTitle(badgeKind) ?: "",
+            formattedRank = "", // TODO: replace with `getBadgeRankName(BadgeRank.LOCKED)`
```

Comment:
What's the plan with the left TODOs?

Reply:
@ivan-magda 
To resolve these TODOs the https://github.com/hyperskill/mobile-app/pull/606 should be merged